File: app/src/main/java/it/unical/mat/embasp/base/Handler.java
package it.unical.mat.embasp.base;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * A collection of {@link InputProgram} and {@link OptionDescriptor}.
 * The subclasses have to implement {@link #startAsync(Callback, List, List)} and {@link #startSync(List, List)} methods.
 * Each instance inside {@link #programs} and {@link #options} are represented by an integer (id) , respectively.
 *
 * @see ArrayList
 * @see HashMap
 * @see List
 * @see Map
 */

public abstract class Handler {
	/**
	 * @param programs
	 *            is where {@link InputProgram} elements are stored.
	 */
	protected Map<Integer, InputProgram> programs;
	/**
	 * @param option
	 *            is where {@link OptionDescriptor} elements are stored
	 */
	protected Map<Integer, OptionDescriptor> options;

	public Handler() {
		programs = new HashMap<>();
		options = new HashMap<>();
	}

	/**
	 * Add a new element inside {@link #options} set.
	 *
	 * @param o
	 *            is the new {@link OptionDescriptor} instance
	 * @return the id associate to the new added {@link OptionDescriptor} instance
	 */

	public int addOption(final OptionDescriptor o) {

		int last_index = options.size();
		final int current_value = last_index;
		options.put(last_index++, o);
		return current_value;
	}

	/**
	 * Add a new element inside {@link #programs} set.
	 *
	 * @param program
	 *            is the {@link InputProgram} instance added to the collection.
	 * @return the id associate to the new added {@link InputProgram} instance.
	 */

	public int addProgram(final InputProgram program) {

		int last_index = programs.size();
		final int current_value = last_index;
		programs.put(last_index++, program);
		return current_value;
	}

	protected List<OptionDescriptor> collect_options(final List<Integer> option_index) {
		final List<OptionDescriptor> input_option = new ArrayList<>();

		if (option_index == null)
			for (final Map.Entry<Integer, OptionDescriptor> option : options.entrySet())
				input_option.add(options.get(option.getKey()));
		else
			for (final int index : option_index)
				input_option.add(options.get(index));
		return input_option;
	}

	protected List<InputProgram> collect_programs(final List<Integer> program_index) {

		final List<InputProgram> input_programs = new ArrayList<>();

		if (program_index == null)
			for (final Map.Entry<Integer, InputProgram> program : programs.entrySet())
				input_programs.add(programs.get(program.getKey()));
		else
			for (final int index : program_index)
				input_programs.add(programs.get(index));
		return input_programs;
	}

	/**
	 * Returns the specified {@link InputProgram} element
	 *
	 * @param key
	 *            the id
	 * @return {@link InputProgram} element associate with the given key
	 */
	public InputProgram getInputProgram(final int key) {
		return programs.get(key);
	}

	/**
	 * Returns the specified {@link OptionDescriptor} element
	 *
	 * @param key
	 *            the id
	 * @return {@link OptionDescriptor} element associate with the given key
	 */
	public OptionDescriptor getOptionDescriptor(final int key) {
		return options.get(key);
	}

	/** Removes all of the elements from {@link #programs} and {@link #options}.Both of the collections will be empty after this method returns */
	public void removeAll() {

		programs.clear();
		options.clear();

	}

	/**
	 * Removes the element associate within the given id from {@link #options} set.
	 *
	 * @param option_id
	 *            the id associate within an element
	 */
	public void removeOption(final int option_id) {

		options.remove(option_id);

	}

	/**
	 * Removes every occurrence of a specified {@link OptionDescriptor} element from {@link #options} set.
	 *
	 * @param o
	 *            element to be removed
	 * @result true if one or more elements are removed , false otherwise
	 */
	public boolean removeOption(final OptionDescriptor o) {

		boolean result = false;
		for (final Map.Entry<Integer, OptionDescriptor> entry : options.entrySet())
			if (entry.getValue().equals(o)) {
				options.remove(entry.getKey());
				result = true;
			}

		return result;

	}

	/**
	 * Removes every occurrence of a specified {@link InputProgram} element from {@link #programs} set.
	 *
	 * @param p
	 *            element to be removed
	 * @result true if one or more elements are removed , false otherwise
	 */

	public boolean removeProgram(final InputProgram p) {

		boolean result = false;
		for (final Map.Entry<Integer, InputProgram> entry : programs.entrySet())
			if (entry.getValue().equals(p)) {
				programs.remove(entry.getKey());
				result = true;
			}
		return result;
	}

	/**
	 * Removes the element associate within the given id from {@link #programs} set.
	 *
	 * @param program_id
	 *            the id associate within an element
	 */

	public void removeProgram(final int program_id) {

		programs.remove(program_id);
	}

	/** @see Service */
	public void startAsync(final Callback c) {
		startAsync(c, (List<Integer>) null, (List<Integer>) null);
	}

	/**
	 * This method have to be implemented by subclasses to execute solver in a asynchronous way, if no parameters are given
	 * the entire sets of programs and option are used
	 */
	public void startAsync(final Callback c, final List<Integer> program_index, final List<Integer> option_index) {
	}

	/** @see #startSync(List, List) */
	public Output startSync() {

		return startSync((List<Integer>) null, (List<Integer>) null);
	}

	/**
	 * This method have to be implemented by subclasses to execute solver in a synchronous way, if no parameters are given
	 * the entire sets of programs and option are used
	 *
	 * @see Service
	 */
	public Output startSync(final List<Integer> program_index, final List<Integer> option_index) {
		return null;
	}
}

File: Python version/base/handler.py
from abc import ABCMeta


class Handler(object):
    """A collection of InputProgram and OptionDescriptor.

    The subclasses have to implement start_async(Callback, List, List)
    and start_sync(List, List) methods.
    """
    __metaclass__ = ABCMeta

    def __init__(self):
        self._programs = dict()  # Is where InputProgram elements are stored.
        self._options = dict()  # Is where OptionDescriptor elements are stored

    def add_option(self, o):
        """Add a new element inside _options dict.

        The o parameter is the new OptionDescriptor instance. The method
        return the id associated whit the added OptionDescriptor
        instance.
        """
        last_index = len(self._options)
        current_value = last_index
        self._options[last_index] = o
        return current_value

    def add_program(self, program):
        """Add a new element inside _programs dict.

        The program param is the InputProgram instance added to the
        collection. The method return the id associated whit the added
        InputProgram instance.
        """
        last_index = len(self._programs)
        current_value = last_index
        self._programs[last_index] = program
        return current_value

    def _collect_options(self, option_index):
        """Return a list of options in _options dict, according to set of
        indexes given.

        If option_index is empty, the method return a list of all
        options.
        """
        input_option = list()
        if not option_index:
            for k in self._options.keys():
                input_option.append(self._options.get(k))
        else:
            for index in option_index:
                input_option.append(self._options.get(index))
        return input_option

    def _collect_programs(self, program_index):
        """Return a list of programs in _programs dict, according to set of
        indexes given.

        If program_index is empty, the method return a list of all
        program.
        """
        input_programs = list()
        if not program_index:
            for k in self._programs.keys():
                input_programs.append(self._programs.get(k))
        else:
            for index in program_index:
                input_programs.append(self._programs.get(index))
        return input_programs

    def get_input_program(self, key):
        """Returns the specified InputProgram element.

        The parameter key represents the id. The method return the
        InputProgram element associated with the given key.
        """
        return self._programs.get(key)

    def get_option_descriptor(self, key):
        """Returns the specified OptionDescriptor element.

        The parameter key represents the id. The method return the
        OptionDescriptor element associated with the given key.
        """
        return self._options.get(key)

    def remove_all(self):
        """Removes all the elements from _programs and _options.

        Both the collections will be empty after this method returns.
        """
        self._options.clear()
        self._programs.clear()

    def remove_option_from_id(self, option_id):
        """Removes the element associated with the given id from _options
        dict.

        option_id represents the id associated with an element.
        """
        self._options.pop(option_id)

    def remove_option_from_value(self, o):
        """Removes every occurrence of a specified OptionDescriptor element
        from _options dict.

        the parameter o represents the element to be removed.
        The method return true if one or more elements are removed,
        false otherwise
        """
        result = False
        for k in self._options:
            if self._options.get(k) == o:
                self._options.pop(k)
                result = True
        return result

    def remove_program_from_value(self, p):
        """Removes every occurrence of a specified InputProgram element from
        _programs dict.

        The parameter p represents the element to be removed.
        The method returns true if one or more elements are removed,
        false otherwise
        """
        result = False
        for k in self._programs:
            if self._programs.get(k) == p:
                self._programs.pop(k)
                result = True
        return result

    def remove_program_from_id(self, program_id):
        """Removes the element associated with the given id from _programs}
        dict.

        The parameter program_id represents the id associated with an
        element
        """
        self._programs.pop(program_id)

    def start_async(self, c, program_index=None, option_index=None):
        """This method have to be implemented by subclasses to execute solver
        in an asynchronous way, if no parameters are given, the entire sets of
        programs and option are used."""
        pass

    def start_sync(self, program_index=None, option_index=None):
        """This method have to be implemented by subclasses to execute solver
        in a synchronous way, if no parameters are given, the entire sets of
        programs and option are used."""
        return None
